### PR TITLE
Make pad more robust to non-hashable values.

### PR DESCRIPTION
**Pull request by jblespiau**
The `pad_width` can be an iterable of iterbles, thus, we need to cast both the inner and outer list to tuples.
Also, the values should be intergers, but some use-cases (mostly Tensorflow) pass e.g. DeviceArray(1) as values in there.

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -2324,10 +2325,12 @@ def _pad(array, pad_width, mode, constant_values):
     msg = "Unimplemented padding mode '{}' for np.pad."
     raise NotImplementedError(msg.format(mode))
 
+
 @_wraps(np.pad)
-def pad(array, pad_width, mode='constant', constant_values=0):
-  if isinstance(pad_width, list):
-    pad_width = tuple(pad_width)
+def pad(array, pad_width, mode="constant", constant_values=0):
```
Comment (jblespiau) [inline jax/_src/numpy/lax_numpy.py:2330]: This is the change, the rest is only identication fixes.

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -2324,10 +2325,12 @@ def _pad(array, pad_width, mode, constant_values):
...
+  pad_width = tuple((tuple(int(i)
```
Comment (hawkinsp) [inline jax/_src/numpy/lax_numpy.py:2331]: Formatting nit: I'd break the line before `if` not before `for`.

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -2324,10 +2325,12 @@ def _pad(array, pad_width, mode, constant_values):
...
+                           for i in x) if not isinstance(x, int) else x
```
Comment (hawkinsp) [inline jax/_src/numpy/lax_numpy.py:2332]: Wouldn't it be simpler just to write:
`tuple(tuple(int(i) for i in x) for x in pad_width)`

Why not unconditionally convert to int?

Comment (hawkinsp) [reply jax/_src/numpy/lax_numpy.py:2332]: Oh, I see, numpy.pad accepts ints and tuples. This is fine.

**Comment by hawkinsp** [comment]
You have some test failures as well.

File: jax/_src/numpy/lax_numpy.py [hunk 4]
```diff
@@ -2324,10 +2325,13 @@ def _pad(array, pad_width, mode, constant_values):
...
     raise NotImplementedError(msg.format(mode))
 
+
 @_wraps(np.pad)
-def pad(array, pad_width, mode='constant', constant_values=0):
-  if isinstance(pad_width, list):
-    pad_width = tuple(pad_width)
+def pad(array, pad_width, mode="constant", constant_values=0):
+  if isinstance(pad_width, Sequence):
+    pad_width = tuple(  # new line
```
Comment (hawkinsp) [inline jax/_src/numpy/lax_numpy.py:2332]: Sometimes I've just done this via a trailing `#` with no text. (Note we don't use pyformat, so you can also just *not* use it here...)

**Review by hawkinsp — APPROVED** [review]

**Pull request merged**